File: editorials/problems/CTD01/solution.c
#include<stdio.h>
#include<string.h>

char* backspace_compare(char *S, char* T) {
	int i=strlen(S)-1, j=strlen(T)-1;
	int skipS=0, skipT=0;

	while (i >= 0 || j >= 0) {
		while (i >= 0) {
			if (S[i] == '#') {skipS++; i--;}
			else if (skipS > 0) {skipS--; i--;}
			else break;
		}
		while (j >= 0) {
			if (T[j] == '#') {skipT++; j--;}
			else if (skipT > 0) {skipT--; j--;}
			else break;
		}
		if (i >= 0 && j >= 0 && S[i] != T[j])
			return "false";
		if ((i >= 0) != (j >= 0))
			return "false";
		i--; j--;
	}
	return "true";
}

int main() {
	char S[1000000], T[1000000];
	int t;
	scanf("%d", &t);
	while(t--) {
		scanf("%s", S);
		scanf("%s", T);
		printf("%s\n", backspace_compare(S, T));
	}
}

File: editorials/problems/CTD01/solution.java
import java.util.*;
class solution {
	public static boolean backspaceCompare(String S, String T) {
		int i = S.length() - 1, j = T.length() - 1;
		int skipS = 0, skipT = 0;

		while (i >= 0 || j >= 0) {
			while (i >= 0) {
				if (S.charAt(i) == '#') {skipS++; i--;}
				else if (skipS > 0) {skipS--; i--;}
				else break;
			}
			while (j >= 0) {
				if (T.charAt(j) == '#') {skipT++; j--;}
				else if (skipT > 0) {skipT--; j--;}
				else break;
			}
			if (i >= 0 && j >= 0 && S.charAt(i) != T.charAt(j))
				return false;
			if ((i >= 0) != (j >= 0))
				return false;
			i--; j--;
		}
		return true;
	}
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int t = scan.nextInt();
		String S, T;
		while(t-- != 0) {
			S = scan.next();
			T = scan.next();
			System.out.println(backspaceCompare(S, T));
		}
		scan.close();
	}
}